### PR TITLE
[AIR] `SklearnTrainer` & `Predictor` interfaces

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR adds interfaces for `SklearnTrainer` and `SklearnPredictor`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        **fit_params: Additional kwargs passed to ``estimator.fit()``
+            method.
+    """
+
+    def __init__(
+        self,
+        *,
+        estimator: BaseEstimator,
+        datasets: Dict[str, GenDataset],
+        label_column: Optional[str] = None,
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:110]: Just to clarify, label is optional to support e.g. unsupervised learning like clustering or so, correct?

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+            model = RandomForestClassifier().fit(train_X, train_y)
+            predictor = SklearnPredictor(model=model)
+
+            data = np.array([[1, 2], [3, 4]])
+            predictions = predictor.predict(data)
+
+            # Only use first and second column as the feature
+            data = np.array([[1, 2, 8], [3, 4, 9]])
+            predictions = predictor.predict(data, feature_columns=[0, 1])
```
Comment (krfricke) [inline python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:76]: I like simple examples, but I'm wondering if we should default t providing end to end AIR examples (e.g. use `SklearnPredictor.from_checkpoint()`). I guess one question is if the constructor of Predictors should actually be private.

**Review by krfricke — APPROVED** [review]
Interface LGTM

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 1 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:110]: Yes!

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py (hunk 2 above)
Comment (Yard1) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:76]: We'll have proper examples in a separate file, this is just a minimum runnable example for the docstring

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @classmethod
+    def from_checkpoint(cls, checkpoint: Checkpoint) -> "SklearnPredictor":
+        """Instantiate the predictor from a Checkpoint.
+
+        The checkpoint is expected to be a result of ``SklearnTrainer``.
+
+        Args:
+            checkpoint (Checkpoint): The checkpoint to load the model and
+                preprocessor from. It is expected to be from the result of a
+                ``SklearnPredictor`` run.
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:36]:
```suggestion
                ``SklearnTrainer`` run.
```

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            from ray.ml.train.integrations.sklearn import SklearnTrainer
+            from sklearn.ensemble import RandomForestRegressor
+
+            train_dataset = ray.data.from_items(
+                [{"x": x, "y": x + 1} for x in range(32)])
+            trainer = SklearnTrainer(
+                sklearn_estimator=RandomForestRegressor,
+                label_column="y",
+                scaling_config={
+                    "trainer_resources": {"CPU": 4}
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:49]: `trainer_resources` is not a valid key in the scaling config right?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+            If ``scoring`` represents multiple scores, one can use:
+
+            - a list or tuple of unique strings;
+            - a callable returning a dictionary where the keys are the metric
+            names and the values are the metric scores;
+            - a dictionary with metric names as keys and callables a values.
+
+        cv: Determines the cross-validation splitting strategy. If specified,
+            cross-validation will be ran on the train dataset, in addition to
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:81]:
```suggestion
            cross-validation will be run on the train dataset, in addition to
```

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            - None, to skip cross-validation.
+            - int, to specify the number of folds in a ``(Stratified)KFold``,
+            - ``CV splitter``,
+            - An iterable yielding (train, test) splits as arrays of indices.
+
+            For int/None inputs, if the estimator is a classifier and ``y`` is
+            either binary or multiclass, ``StratifiedKFold`` is used. In all
+            other cases, ``KFold`` is used. These splitters are instantiated
+            with ``shuffle=False`` so the splits will be the same across calls.
+        scaling_config: Configuration for how to scale data parallel training.
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:96]: This won't be doing data parallel training, is that right?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            - a callable that returns a single value.
+
+            If ``scoring`` represents multiple scores, one can use:
+
+            - a list or tuple of unique strings;
+            - a callable returning a dictionary where the keys are the metric
+            names and the values are the metric scores;
+            - a dictionary with metric names as keys and callables a values.
+
+        cv: Determines the cross-validation splitting strategy. If specified,
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:80]: How does this fit with the cross validation in Tuner cc @xwjiang2010?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+ScoringType = Union[MetricType, Iterable[MetricType], Dict[str, MetricType]]
+CVType = Union[int, Iterable, BaseCrossValidator]
+
+
+@PublicAPI(stability="alpha")
+class SklearnTrainer(Trainer):
+    """A Trainer for scikit-learn estimator training.
+
+    This Trainer runs the ``fit`` method of the given estimator in a
+    non-distributed manner on a single Ray Actor.
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:30]: Is the plan to eventually have this use our Ray+joblib integration?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    The ``n_jobs`` (or ``thread_count``) parameters will be set to match
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: For my own understanding, would multithreading even be effective here if we have python GIL?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 4 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:49]: It should be? It's in the dataclass

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 6 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:96]: No, it's just for validation. It will all be ran on a single actor.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 7 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:80]: We can switch to any distributed CV we have later. ATM this will not be distributed

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 6 above)
Comment (amogkam) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:96]: Should we then change the docstring to not mention data parallel training?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 4 above)
Comment (amogkam) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:49]: oops sorry, yes you're right- forgot that we had changed this!

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 8 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:30]: Yes, but it is a bit tricky. Not all sklearn estimators use joblib. Some can have only single-node parallelism, and there is no way to detect which uses what. If we set the trainable to use eg. 4 CPUs, then the joblib spawned Actors will not be able to be scheduled. We can perhaps leave it up to the user.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 9 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: Most of the time, this is either joblib or C-level parallelism (independent of GIL).

Comment (amogkam) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: From what I understand the default joblib parallelism uses multiprocessing, is that correct? In which case I don't think multiprocessing will work well inside the trainable/Ray actor.

Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: yes, that is correct. we will have to force a thread joblib backend for now. We will want to use the Ray one eventually, but I don't know yet how to solve the issue with resources (see conversation above).

Comment (amogkam) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: Got it yeah- but what I'm asking is for thread joblib, from what I'm reading online, it seems like it's useful if you're able to disable the GIL. Is that something that we have to do, or that the user will need to handle in their code?

Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: I don't think we can disable the GIL, right?

The bottom line here is that we will be able to make estimators that make use of non-joblib parallelism use the correct number of CPUs, but in the initial implementation, I don't anticipate any real parallelism for joblib-based estimators.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 6 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:96]: Yes!

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            - None, to skip cross-validation.
+            - int, to specify the number of folds in a ``(Stratified)KFold``,
+            - ``CV splitter``,
+            - An iterable yielding (train, test) splits as arrays of indices.
+
+            For int/None inputs, if the estimator is a classifier and ``y`` is
+            either binary or multiclass, ``StratifiedKFold`` is used. In all
+            other cases, ``KFold`` is used. These splitters are instantiated
+            with ``shuffle=False`` so the splits will be the same across calls.
+        scaling_config: Configuration for how to scale data parallel training.
```
Comment (Yard1) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:96]:
```suggestion
        scaling_config: Configuration for how to scale training.
```

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 9 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: I'll think about how to make ray joblib work here, though.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+            from ray.ml.train.integrations.sklearn import SklearnTrainer
+            from sklearn.ensemble import RandomForestRegressor
+
+            train_dataset = ray.data.from_items(
+                [{"x": x, "y": x + 1} for x in range(32)])
+            trainer = SklearnTrainer(
+                sklearn_estimator=RandomForestRegressor,
+                label_column="y",
+                scaling_config={
```
Comment (ericl) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:48]: Do we raise a good error message is specifying unsupported resources configs, like num_workers? It would be great to have unit tests for these cases.

Comment (ericl) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:48]: Maybe trainers should have a class attribute that lists supported resources, and the base class can auto raise unsupported config exceptions, for uniformity across trainers.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 9 above)
Comment (ericl) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:32]: +1 on joblib as future work. Maybe leave this as a TODO that true multi-processing is possible but not implemented yet? That clarifies to the user why this isn't offered.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 11 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:48]: I was planning on having just a check in this trainer, but making it more generic is a good idea. I will do that during implementation.
